### PR TITLE
Connect SQL verifier tag transforms to policy rules in datalog.

### PR DESCRIPTION
**Pull request by markww**
This change connects the SQL verifier tag transform nodes to the policy
rules that control their behavior, and shows this behavior control
occurring in unit tests.

### CODE REVIEW

File: src/analysis/souffle/tag_transforms.dl [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+#include "src/analysis/souffle/operations.dl"
+#include "src/analysis/souffle/tags.dl"
+
+// Note: this will need to change when we integrate auth logic into the SQL
+// verifier.
+#define DEFAULT_SQL_OWNER "sql"
+#define TAG_TRANSFORM_OPERATOR "sql.tag_transform"
+#define TAG_TRANSFORM_RULE_NAME_ATTR "rule_name"
+
+.type SqlPolicyRulePrecondition = [ preconditionName : symbol, requiredTag: IntegrityTag ]
```
Comment (bgogul) [inline src/analysis/souffle/tag_transforms.dl:28]: Mostly a clarification: is the `preconditionName` the name `a` and `b` from a pattern such as `a? + b?`?

File: src/analysis/souffle/tag_transforms.dl [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+  operationHasAttribute(op, [TAG_TRANSFORM_RULE_NAME_ATTR, $StringAttributePayload(ruleName)]).
+
+.decl tagTransformHasResultAccessPath(tagTransform: Operation, resultPath: AccessPath)
+tagTransformHasResultAccessPath([owner, operator, result, operandList, attrs], result) :-
+  isTagTransformOperation([owner, operator, result, operandList, attrs]).
+
+.decl tagTransformHasIndexedTagPrecondition(
+  tagTransform: Operation, index: number, tagPrecondition: TagPrecondition)
+
+tagTransformHasIndexedTagPrecondition(op, index, [preconditionOperand, requiredTag]) :-
```
Comment (bgogul) [inline src/analysis/souffle/tag_transforms.dl:106]: is `index` needed at all? Don't we just care about `preconditionOperand`?

Comment (bgogul) [reply src/analysis/souffle/tag_transforms.dl:106]: Ah, the index is mainly needed to implement `forall` as used in `allTagTransformPreconditionsLessThanIndexMet`? Perhaps, document the reason somewhere?

**Review by bgogul — APPROVED** [review]
Nice! ADTs do make the code easy to read.

File: src/analysis/souffle/BUILD [hunk 3]
```diff
@@ -64,6 +81,28 @@ souffle_cc_binary(
...
+    included_dl_scripts = [
+        "attributes.dl",
+        "authorization_logic.dl",
+        "check_predicate.dl",
+        "dataflow_graph.dl",
+        "operations.dl",
+        "tag_transforms.dl",
+        "tags.dl",
+        "taint.dl",
+    ],
```
[the remark this answers is not in the document]
Comment (markww) [reply src/analysis/souffle/BUILD:97]: I know, right? I started a draft CL at one point to attempt to use a `filegroup`, but I ran into issues that I do not clearly remember.

[the remark this answers is not in the document]
Comment (markww) [reply src/analysis/souffle/BUILD:97]: We could just run the preprocessor on `taint.dl` to produce one monster header that we can include elsewhere. A bit crude, but could work if we're careful.

File: src/analysis/souffle/tag_transforms.dl (hunk 1 above)
Comment (markww) [reply src/analysis/souffle/tag_transforms.dl:28]: Yes, that is correct.

File: src/analysis/souffle/check_predicate.dl [hunk 4]
```diff
@@ -22,8 +22,10 @@
 
 .type Bool <: number
 
+.type IOrCTag = ITag{ iTag: IntegrityTag } | CTag{ cTag: Tag }
```
[the remark this answers is not in the document]
Comment (markww) [reply src/analysis/souffle/check_predicate.dl]: Done.

File: src/analysis/souffle/tag_transforms.dl (hunk 2 above)
Comment (markww) [reply src/analysis/souffle/tag_transforms.dl:111]: Added a comment explaining this.

File: src/analysis/souffle/tag_transforms.dl (hunk 1 above)
Comment (markww) [reply src/analysis/souffle/tag_transforms.dl:31]: Added a comment explaining this.

**Pull request closed without merging**